### PR TITLE
Color conversion - optimizations and tests

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**Bug fix**
**Refactoring**
**Tests**

## Description:
The precision of RGB<->YCbCr conversions is improved. They are now closer to being fully reversible.
It also adds unit tests to these functions.
Refactoring: change namespaces to classes to facilitate testing.
Fixed a bug in conversion Y<->Gray

### Affected modules and functionalities:
Color space conversion, JPEG distortion

### Tests:

- [ ] Existing tests apply
- [X] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_impl.cuh [hunk 1]
```diff
@@ -98,18 +98,20 @@ template <bool horz_subsample, bool vert_subsample, typename T>
 __inline__ __device__
 void ycbcr_to_rgb_subsampled(ivec2 offset, const Surface2D<uint8_t>& out,
                              YCbCrSubsampled<T, horz_subsample, vert_subsample> ycbcr) {
-  using kernels::color::jpeg::ycbcr_to_rgb;
+  auto ycbcr_to_rgb = [](auto x) {
```
Comment (jantonguirao) [inline dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_impl.cuh:101]: why this change?

Comment (mzient) [reply dali/kernels/imgproc/jpeg/jpeg_distortion_gpu_impl.cuh:101]: Because it's no longer a namespace, so "using" doesn't work.

**Comment by mzient** [comment]
!build

**Comment by stiepan** [comment]
Description nitpick: Could you point out what bug is being fixed here? Was it the poor precision or something else?

**Comment by mzient** [comment]
> Description nitpick: Could you point out what bug is being fixed here? Was it the poor precision or something else?

There was a bug in conversion Y<->Gray.
Updated the description.

**Review by jantonguirao — APPROVED** [review]

File: dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h [hunk 2]
```diff
@@ -39,172 +39,263 @@ DALI_HOST_DEV DALI_FORCEINLINE vec<N, float> norm(vec<N, float> x) {
...
+// Y, Cb, Cr definition from ITU-R BT.601, with values in the range 16-235, allowing for
+// footroom and headroom
+struct itu_r_bt_601 {
+  template <typename Output, typename Input>
+  static DALI_HOST_DEV DALI_FORCEINLINE constexpr Output rgb_to_y(vec<3, Input> rgb) {
+    constexpr vec3 coeffs = vec3(0.25678823529f, 0.50412941176f, 0.09790588235f)
+                            * detail::scale_factor<Input, Output>();
+    constexpr float bias = 0.0625f * detail::bias_scale<Output>();
+    float y = dot(coeffs, rgb) + bias;
+    return needs_clamp<Input, Output>::value ? Convert<Output>(y) : ConvertSat<Output>(y);
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h:84]: It would be a very weird function, because it doesn't convert from Input, but from an unnormalized float. - normally, ConvertSat doesn't clamp when it doesn't have to. Here we know that the intermediate float value is in range.

File: dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h [hunk 3]
```diff
@@ -39,172 +39,263 @@ DALI_HOST_DEV DALI_FORCEINLINE vec<N, float> norm(vec<N, float> x) {
...
-  constexpr vec3 coeffs(0.299f, 0.587f, 0.114f);
-  auto rgb = detail::norm(rgb_in);  // TODO(janton): optimize number of multiplications
-  return ConvertSatNorm<Output>(dot(coeffs, rgb));
-}
 
 template <>
-DALI_HOST_DEV DALI_FORCEINLINE uint8_t rgb_to_y(vec<3, uint8_t> rgb) {
-  constexpr vec3 coeffs(0.299f, 0.587f, 0.114f);
+DALI_HOST_DEV DALI_FORCEINLINE
+uint8_t jpeg::rgb_to_y<uint8_t, uint8_t>(vec<3, uint8_t> rgb) {
```
Comment (stiepan) [inline dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h:268]: Is this specialization necessary?

Comment (mzient) [reply dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h:268]: Perhaps not.

Comment (stiepan) [reply dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h:268]: And are we good with keeping the unnecessary ones? A downside is that it may be consufing for any code readers as to why given case needs special treatment.

Comment (mzient) [reply dali/kernels/imgproc/color_manipulation/color_space_conversion_impl.h]: Removed.

**Review by stiepan — APPROVED** [review]
